### PR TITLE
fix: some bug fix about publish page

### PR DESCRIPTION
**Pull request by alanlong9278**
### bugfix
- can not get all bot publish histories. It just get the first bot publish history when initial the page
-  if the previous publish failed or success, start a new publish process, the notification modal will pop up the old ones.  The notification card display wrong when publish bot again. It will show the old notification card when publishing again

## Task Item

#minor

## Screenshots

notification
image

### CODE REVIEW

File: Composer/packages/client/src/pages/publish/Publish.tsx [hunk 1]
```diff
@@ -210,24 +219,28 @@ const Publish: React.FC<RouteComponentProps<{ projectId: string; targetName?: st
     }
   }, [projectId]);
 
+  const usePrevious = (value) => {
+    const ref = useRef();
```
Comment (a-b-r-o-w-n) [inline Composer/packages/client/src/pages/publish/Publish.tsx:223]: I think this violates react rules on using hooks. They need to be used deterministically in the render of the component.

**Review by a-b-r-o-w-n — CHANGES_REQUESTED** [review]

File: Composer/packages/client/src/pages/publish/Publish.tsx (hunk 1 above)
Comment (alanlong9278) [reply Composer/packages/client/src/pages/publish/Publish.tsx:223]: Move usePrevious out of component

File: Composer/packages/client/src/pages/publish/Publish.tsx [hunk 2]
```diff
@@ -210,24 +226,20 @@ const Publish: React.FC<RouteComponentProps<{ projectId: string; targetName?: st
     }
   }, [projectId]);
 
+  const prevBotStatusList: IBotStatus[] | undefined = usePrevious(botStatusList);
   useEffect(() => {
-    // init publishHistoryList
-    if (Object.keys(publishHistoryList).length > 0) {
+    // init bot status list for the botProjectData is empty array when first mounted
+    if (!isEqual(prevBotStatusList, statusList)) {
```
Comment (a-b-r-o-w-n) [inline Composer/packages/client/src/pages/publish/Publish.tsx:232]: I am having a hard time understanding why this is necessary.

This is only used to compare the status list and then update if changed. Can that not be done using the current state value?

```ts
if (!isEqual(currentStatusList, statusList) {
  setCurrentStatusList(statusList)
}
```

The `usePrevious` hook updates the ref in an effect, which is async anyways so this value could be stale.

**Review by yeze322 — CHANGES_REQUESTED** [review]
Several advices to make sure readability:
- Avoid nested if else clauses, return early.
- Extract complicated state computation logic as pure functions.
- Define necessary local variables to replace appearances of `botPublishHistory[0].status`. Avoid context switching when reading

File: Composer/packages/client/src/pages/publish/Publish.tsx (hunk 2 above)
Comment (alanlong9278) [reply Composer/packages/client/src/pages/publish/Publish.tsx]: Remove `useRef` hook.

**Comment by boydc2014** [comment]
I agree we can merge it if this is functional working. But a few more comments for refactoring. See below

File: Composer/packages/client/src/pages/publish/Publish.tsx [hunk 3]
```diff
@@ -159,49 +159,85 @@ const Publish: React.FC<RouteComponentProps<{ projectId: string; targetName?: st
   };
 
   const [pendingNotification, setPendingNotification] = useState<Notification>();
+  const [previousBotPublishHistoryList, setPreviousBotPublishHistoryList] = useState(botPublishHistoryList);
```
Comment (boydc2014) [inline Composer/packages/client/src/pages/publish/Publish.tsx:162]: can we have a better naming for this what previousHistory means? do we have a current history?

and there is no good to add "list" as postfix, botPublishHistory is good enough

File: Composer/packages/client/src/pages/publish/Publish.tsx [hunk 4]
```diff
@@ -159,49 +159,85 @@ const Publish: React.FC<RouteComponentProps<{ projectId: string; targetName?: st
...
-                if (item.id === botProjectId) {
-                  item.status = botPublishHistory[0].status;
-                  item.comment = botPublishHistory[0].comment;
-                  item.message = botPublishHistory[0].message;
-                  item.time = botPublishHistory[0].time;
-                }
-                return item;
-              })
-            );
+      if (!(bot.publishTarget && bot.publishTargets)) {
```
Comment (boydc2014) [inline Composer/packages/client/src/pages/publish/Publish.tsx:167]: First, the comments "// most recent item is a 202, which means we should poll for updates..." is not related to this line.  there is even no 202 in many lines. 

For such a long use effect and the many other long use effects below, we should give a top-level summary.

**Review by boydc2014 — APPROVED** [review]

**Pull request merged**